Ingress egress filters

dpino opened the pull request:
Early implementation of ingress/egress filters. 
- Incoming packets from inet and b4 are filtered by an ingress filter, same applies for outgoing packets with an egress filter.
- Filters are configured in a lwaftr.conf file. Only one rule per type of filter is allowed. I think in case of needing more than one rule, that can be easily done using pflang language (concat rules wiith 'or' or 'and').
- As for egress filter, they also applied to ICMP packets sent in a ICMP policy. I'm really not sure about this. I think if a ICMP policy is enabled those packets should not be filtered by an egress filter.

dpino committed:
Add ingress/egress filter tests to end-to-end.sh

dpino committed:
Add ingress/egress filter tests to end-to-end-vlan.sh

dpino committed:
Implement ingress/egress filtering

dpino commented:
Rebased and updated PR.

kbara commented:
I think the ingress and egress rules should be in separate files. The conf files are already getting cluttered and not composing well for our tests as we add new things. The general concept looks good, though some thought needs to be given to what happens when a rule in a bigger chain fails (ie, due to a packet access check being out of bounds).

dpino committed:
Compile filters from ACL file

Filters are stored in an ACL file. Format:

ingress_filter_ipv4,ingress_filter_ipv6,egress_filter_ipv4,egress_filter_ipv6
rule_11,rule_21,rule_31,rule_41
rule_12,rule_22,,rule_42

One single rule is composed for each type of filter appending the column
rules with an 'and' keyword. Example:

ingress_filter_ipv4 = rule_11 and rule_12

Each pflang rule is compiled with pflua to a Lua function which are
exectuted later when a packet arrives or is sent.

dpino commented:
Rebased PR. Filters are stored in an ACL file. There are 4 filters in total: ingress_filter_ipv4, ingress_filter_ipv6, egress_filter_ipv4, egress_filter_ipv6.

Each column is a type of filter. The rules for a filter are composed appending all the values of a column using the 'and' keyword. The final string is a filter in pflang language which is compiled to a Lua function via pflua. The compiled filters are stored as properties in the conf file.

Whenever a packet is received or transmited, filters are executed if defined.

dpino commented:
It seems is failing due to not being able to locate the right path for the ACL files.

dpino committed:
Move test filters to independent file

Travis is unable to locate proper path to ACL file.

dpino commented:
This PR solves #147

andywingo commented on src/apps/lwaftr/acl.lua:
```diff
@@ -0,0 +1,76 @@
+   }
+end
+
+function compile(filename, args)
```
This function needs a comment about the grammar of the file and what the arguments mean.  As it, I can't really tell what's going on.  I also wonder why you feel the need to index filters by named number instead of by name (i.e. foo.egress_fitler_ipv4 vs foo[EGRESS_FILTER_IPV4], but that could make sense to me if there were more comments :)

dpino committed:
Add inline documentation about ACL file format

Removed named indexes for filters.

dpino commented:
Removed named indexes for filters and added inline documentation about ACL file format.

andywingo commented on src/program/snabb_lwaftr/tests/data/no_icmp_with_filters_accept.conf:
```diff
@@ -0,0 +1,18 @@
+policy_icmpv4_outgoing = policies['DROP'],
+policy_icmpv6_outgoing = policies['DROP'],
+vlan_tagging = false,
+acl = 'accept_acl.csv',
```
Missing this file

andywingo commented on src/program/snabb_lwaftr/tests/data/no_icmp_with_filters_drop.conf:
```diff
@@ -0,0 +1,18 @@
+policy_icmpv4_outgoing = policies['DROP'],
+policy_icmpv6_outgoing = policies['DROP'],
+vlan_tagging = false,
+acl = 'drop_acl.csv',
```
Missing this file as well.  How was the CI passing?

andywingo commented on src/apps/lwaftr/acl.lua:
```diff
@@ -0,0 +1,87 @@
+   ipv4_ingress_filter,ipv6_ingress_filter,ipv4_egress_filter,ipv6_egress_filter
+   rule11,rule21,,rule41
+   rule12,rules22,,
+   rule13,,,
```
I'm not sure this format buys us much.  The intention of specifying a CSV file was that we could do performance testing over a more minimal range of ingress/egress filters, but if we're just ANDing together a bunch of rules, why not specify them directly as pflang?  And if not, what are we doing?  Dunno.  Also, will any of this work with vlans?  That is a separate work item that we should schedule appropriately :)

andywingo replied:
It could indeed be that pflang is the right way to go, for what it's worth.  If that's the case, then we could simply replace any thought of CSV files with conf parameters directly, i.e. ipv4_ingress_filter = ...  Let's talk about it tomorrow.

dpino committed:
Add ACL files

dpino committed:
Append filters tests to end-to-end.sh and end-to-end-vlan.sh

dpino committed:
Remove end-to-end-filters.sh

dpino commented:
Outdated by #207
